File: Code_Chapter 3/Code_Chapter 3/Code/Python/Python/SparkSQL.py

# coding: utf-8

# In[1]:

from pyspark.sql import Row
# Creation of the list from where the RDD is going to be created
acTransList = ["SB10001,1000", "SB10002,1200", "SB10003,8000", "SB10004,400", "SB10005,300", "SB10006,10000", "SB10007,500", "SB10008,56", "SB10009,30","SB10010,7000", "CR10001,7000", "SB10002,-10"]
# Create the DataFrame
acTransDF = sc.parallelize(acTransList).map(lambda trans: trans.split(",")).map(lambda p: Row(accNo=p[0], tranAmount=float(p[1]))).toDF()
# Register temporary table in the DataFrame for using it in SQL
acTransDF.createOrReplaceTempView("trans")


# In[2]:

# Print the structure of the DataFrame
acTransDF.printSchema()


# In[3]:

# Show the first few records of the DataFrame
acTransDF.show()


# In[4]:

# Use SQL to create another DataFrame containing the good transaction records
goodTransRecords = spark.sql("SELECT accNo, tranAmount FROM trans WHERE accNo like 'SB%' AND tranAmount > 0")
# Register temporary table in the DataFrame for using it in SQL
goodTransRecords.createOrReplaceTempView("goodtrans")
# Show the first few records of the DataFrame
goodTransRecords.show()


# In[5]:

# Use SQL to create another DataFrame containing the high value transaction records
highValueTransRecords = spark.sql("SELECT accNo, tranAmount FROM goodtrans WHERE tranAmount > 1000")
# Show the first few records of the DataFrame
highValueTransRecords.show()


# In[6]:

# Use SQL to create another DataFrame containing the bad account records
badAccountRecords = spark.sql("SELECT accNo, tranAmount FROM trans WHERE accNo NOT like 'SB%'")
# Show the first few records of the DataFrame
badAccountRecords.show()


# In[7]:

# Use SQL to create another DataFrame containing the bad amount records
badAmountRecords = spark.sql("SELECT accNo, tranAmount FROM trans WHERE tranAmount < 0")
# Show the first few records of the DataFrame
badAmountRecords.show()


# In[8]:

# Do the union of two DataFrames and create another DataFrame
badTransRecords = badAccountRecords.union(badAmountRecords)
# Show the first few records of the DataFrame
badTransRecords.show()


# In[9]:

# Calculate the sum
sumAmount = spark.sql("SELECT sum(tranAmount)as sum FROM goodtrans")
# Show the first few records of the DataFrame
sumAmount.show()


# In[10]:

# Calculate the maximum
maxAmount = spark.sql("SELECT max(tranAmount) as max FROM goodtrans")
# Show the first few records of the DataFrame
maxAmount.show()


# In[11]:

# Calculate the minimum
minAmount = spark.sql("SELECT min(tranAmount)as min FROM goodtrans")
# Show the first few records of the DataFrame
minAmount.show()


# In[12]:

# Use SQL to create another DataFrame containing the good account numbers
goodAccNos = spark.sql("SELECT DISTINCT accNo FROM trans WHERE accNo like 'SB%' ORDER BY accNo")
# Show the first few records of the DataFrame
goodAccNos.show()


# In[13]:

# Calculate the sum using mixing of DataFrame and RDD like operations
sumAmountByMixing = goodTransRecords.rdd.map(lambda trans: trans.tranAmount).reduce(lambda a,b : a+b)
sumAmountByMixing


# In[14]:

# Calculate the maximum using mixing of DataFrame and RDD like operations
maxAmountByMixing = goodTransRecords.rdd.map(lambda trans: trans.tranAmount).reduce(lambda a,b : a if a > b else b)
maxAmountByMixing


# In[15]:

# Calculate the minimum using mixing of DataFrame and RDD like operations
minAmountByMixing = goodTransRecords.rdd.map(lambda trans: trans.tranAmount).reduce(lambda a,b : a if a < b else b)
minAmountByMixing


# In[16]:

# Show the first few records of the DataFrame
acTransDF.show()


# In[17]:

# Print the structure of the DataFrame
acTransDF.printSchema()


# In[18]:

# Create the DataFrame using API for the good transaction records
goodTransRecords = acTransDF.filter("accNo like 'SB%'").filter("tranAmount > 0")
# Show the first few records of the DataFrame
goodTransRecords.show()


# In[19]:

# Create the DataFrame using API for the high value transaction records
highValueTransRecords = goodTransRecords.filter("tranAmount > 1000")
# Show the first few records of the DataFrame
highValueTransRecords.show()


# In[20]:

# Create the DataFrame using API for the bad account records
badAccountRecords = acTransDF.filter("accNo NOT like 'SB%'")
# Show the first few records of the DataFrame
badAccountRecords.show()


# In[21]:

# Create the DataFrame using API for the bad amount records
badAmountRecords = acTransDF.filter("tranAmount < 0")
# Show the first few records of the DataFrame
badAmountRecords.show()


# In[22]:

# Do the union of two DataFrames and create another DataFrame
badTransRecords = badAccountRecords.union(badAmountRecords)
# Show the first few records of the DataFrame
badTransRecords.show()


# In[23]:

# Calculate the sum
sumAmount = goodTransRecords.agg({"tranAmount": "sum"})
# Show the first few records of the DataFrame
sumAmount.show()


# In[24]:

# Calculate the maximum
maxAmount = goodTransRecords.agg({"tranAmount": "max"})
# Show the first few records of the DataFrame
maxAmount.show()


# In[25]:

# Calculate the minimum
minAmount = goodTransRecords.agg({"tranAmount": "min"})
# Show the first few records of the DataFrame
minAmount.show()


# In[26]:

# Create the DataFrame using API for the good account numbers
goodAccNos = acTransDF.filter("accNo like 'SB%'").select("accNo").distinct().orderBy("accNo")
# Show the first few records of the DataFrame
goodAccNos.show()


# In[27]:

# Persist the data of the DataFrame into a Parquet file
acTransDF.write.parquet("python.trans.parquet")


# In[28]:

# Read the data into a DataFrame from the Parquet file
acTransDFfromParquet = spark.read.parquet("python.trans.parquet")


# In[29]:

# Show the first few records of the DataFrame
acTransDFfromParquet.show()


# In[30]:

from pyspark.sql import Row


# In[31]:

# Creation of the list from where the RDD is going to be created
acTransList = ["SB10001,1000", "SB10002,1200", "SB10001,8000","SB10002,400", "SB10003,300", "SB10001,10000","SB10004,500","SB10005,56","SB10003,30","SB10002,7000", "SB10001,-100","SB10002,-10"]
# Create the DataFrame
acTransDF = sc.parallelize(acTransList).map(lambda trans: trans.split(",")).map(lambda p: Row(accNo=p[0], tranAmount=float(p[1]))).toDF()
# Register temporary table in the DataFrame for using it in SQL
acTransDF.createOrReplaceTempView("trans")
# Use SQL to create another DataFrame containing the account summary records
acSummary = spark.sql("SELECT accNo, sum(tranAmount) as transTotal FROM trans GROUP BY accNo")
# Show the first few records of the DataFrame
acSummary.show()    


# In[32]:

# Create the DataFrame using API for the account summary records
acSummaryViaDFAPI = acTransDF.groupBy("accNo").agg({"tranAmount": "sum"}).selectExpr("accNo", "`sum(tranAmount)` as transTotal")
# Show the first few records of the DataFrame
acSummaryViaDFAPI.show()


# In[33]:

# Creation of the list from where the RDD is going to be created
AcMaster = Row('accNo', 'firstName', 'lastName')
AcBal = Row('accNo', 'balanceAmount')
acMasterList = ["SB10001,Roger,Federer","SB10002,Pete,Sampras", "SB10003,Rafael,Nadal","SB10004,Boris,Becker", "SB10005,Ivan,Lendl"]
acBalList = ["SB10001,50000", "SB10002,12000","SB10003,3000", "SB10004,8500", "SB10005,5000"]
# Create the DataFrame
acMasterDF = sc.parallelize(acMasterList).map(lambda trans: trans.split(",")).map(lambda r: AcMaster(*r)).toDF()
acBalDF = sc.parallelize(acBalList).map(lambda trans: trans.split(",")).map(lambda r: AcBal(r[0], float(r[1]))).toDF()
# Persist the data of the DataFrame into a Parquet file
acMasterDF.write.parquet("python.master.parquet")
# Persist the data of the DataFrame into a JSON file
acBalDF.write.json("pythonMaster.json")
# Read the data into a DataFrame from the Parquet file
acMasterDFFromFile = spark.read.parquet("python.master.parquet")
# Register temporary table in the DataFrame for using it in SQL
acMasterDFFromFile.createOrReplaceTempView("master")
# Register temporary table in the DataFrame for using it in SQL
acBalDFFromFile = spark.read.json("pythonMaster.json")
# Register temporary table in the DataFrame for using it in SQL
acBalDFFromFile.createOrReplaceTempView("balance")
# Show the first few records of the DataFrame
acMasterDFFromFile.show()


# In[34]:

# Show the first few records of the DataFrame
acBalDFFromFile.show()


# In[35]:

# Use SQL to create another DataFrame containing the account detail records
acDetail = spark.sql("SELECT master.accNo, firstName, lastName, balanceAmount FROM master, balance WHERE master.accNo = balance.accNo ORDER BY balanceAmount DESC")
# Show the first few records of the DataFrame
acDetail.show()


# In[36]:

# Create the DataFrame using API for the account detail records
acDetailFromAPI = acMasterDFFromFile.join(acBalDFFromFile, acMasterDFFromFile.accNo == acBalDFFromFile.accNo).sort(acBalDFFromFile.balanceAmount, ascending=False).select(acMasterDFFromFile.accNo, acMasterDFFromFile.firstName, acMasterDFFromFile.lastName, acBalDFFromFile.balanceAmount)
# Show the first few records of the DataFrame
acDetailFromAPI.show()


# In[37]:

# Use SQL to create another DataFrame containing the top 3 account detail records
acDetailTop3 = spark.sql("SELECT master.accNo, firstName, lastName, balanceAmount FROM master, balance WHERE master.accNo = balance.accNo ORDER BY balanceAmount DESC").limit(3)
# Show the first few records of the DataFrame
acDetailTop3.show()


# In[ ]:

File: Code_Chapter 4/CodeAndData/R/SparkSQL.R
# SPARK_HOME settings and SparkR library path settings
# When you are running this script from your setup, please make sure that it is pointing to your SPARK_HOME directory
SPARK_HOME_DIR <- "/Users/RajT/source-code/spark-source/spark-2.0"
DATA_DIR <- "/Users/RajT/Documents/Writing/SparkForBeginners/To-PACKTPUB/Contents/B05289-04-SparkProgrammingWithR/CodeAndData/R/"
Sys.setenv(SPARK_HOME=SPARK_HOME_DIR)
.libPaths(c(file.path(Sys.getenv("SPARK_HOME"), "R", "lib"), .libPaths()))
library(SparkR)
spark <- sparkR.session(master="local[*]")

#Fundamentals of R
# Variable
x <- 5
x
# Numerical vector
aNumericVector <- c(10,10.5,31.2,100)
aNumericVector
# Character vector
aCharVector <- c("apple", "orange", "mango")
aCharVector
# Boolean Vector
aBooleanVector <- c(TRUE, FALSE, TRUE, FALSE, FALSE)
aBooleanVector
# List containing multiple vectors
aList <- list(aNumericVector, aCharVector)
aList
# Matrix creation
aMatrix <- matrix(c(100, 210, 76, 65, 34, 45),nrow=3,ncol=2,byrow = TRUE)
aMatrix
bMatrix <- matrix(c(100, 210, 76, 65, 34, 45),nrow=3,ncol=2,byrow = FALSE)
bMatrix
# Create R DataFrame using vectors
ageVector <- c(21, 35, 52) 
nameVector <- c("Thomas", "Mathew", "John")
marriedVector <- c(FALSE, TRUE, TRUE)
aDataFrame <- data.frame(ageVector, nameVector, marriedVector)
colnames(aDataFrame) <- c("Age","Name", "Married")
# DIsplay the contents from the DataFrame
head(aDataFrame)
tail(aDataFrame)
nrow(aDataFrame)
ncol(aDataFrame)
aDataFrame[1]
aDataFrame[2]
aDataFrame[c("Age", "Name")]
aDataFrame[[2]]
aDataFrame[2,]
aDataFrame[c(1,2),]

#Spark DataFrame programming with R - Basics
# Read data from a JSON file to create DataFrame
acTransDF <- read.json(paste(DATA_DIR, "TransList1.json", sep = ""))
# Print the structure of the DataFrame
print(acTransDF)
# Show sample records from the DataFrame
showDF(acTransDF)
# Register temporary view definition in the DataFrame for SQL queries
createOrReplaceTempView(acTransDF, "trans")
# DataFrame containing good transaction records using SQL
goodTransRecords <- sql("SELECT AccNo, TranAmount FROM trans WHERE AccNo like 'SB%' AND TranAmount > 0")
# Register temporary table definition in the DataFrame for SQL queries
createOrReplaceTempView(goodTransRecords, "goodtrans")
# Show sample records from the DataFrame
showDF(goodTransRecords)
# DataFrame containing high value transaction records using SQL
highValueTransRecords <- sql("SELECT AccNo, TranAmount FROM goodtrans WHERE TranAmount > 1000")
# Show sample records from the DataFrame
showDF(highValueTransRecords)
# DataFrame containing bad account records using SQL
badAccountRecords <- sql("SELECT AccNo, TranAmount FROM trans WHERE AccNo NOT like 'SB%'")
# Show sample records from the DataFrame
showDF(badAccountRecords)
# DataFrame containing bad amount records using SQL
badAmountRecords <- sql("SELECT AccNo, TranAmount FROM trans WHERE TranAmount < 0")
# Show sample records from the DataFrame
showDF(badAmountRecords)
# Create a DataFrame by taking the union of two DataFrames
badTransRecords <- union(badAccountRecords, badAmountRecords)
# Show sample records from the DataFrame
showDF(badTransRecords)
# DataFrame containing sum amount using SQL
sumAmount <- sql("SELECT sum(TranAmount) as sum FROM goodtrans")
# Show sample records from the DataFrame
showDF(sumAmount)
# DataFrame containing maximum amount using SQL
maxAmount <- sql("SELECT max(TranAmount) as max FROM goodtrans")
# Show sample records from the DataFrame
showDF(maxAmount)
# DataFrame containing minimum amount using SQL
minAmount <- sql("SELECT min(TranAmount)as min FROM goodtrans")
# Show sample records from the DataFrame
showDF(minAmount)
# DataFrame containing good account number records using SQL
goodAccNos <- sql("SELECT DISTINCT AccNo FROM trans WHERE AccNo like 'SB%' ORDER BY AccNo")
# Show sample records from the DataFrame
showDF(goodAccNos)
# DataFrame containing good transaction records using API
goodTransRecordsFromAPI <- filter(acTransDF, "AccNo like 'SB%' AND TranAmount > 0")
# Show sample records from the DataFrame
showDF(goodTransRecordsFromAPI)
# DataFrame containing high value transaction records using API
highValueTransRecordsFromAPI = filter(goodTransRecordsFromAPI, "TranAmount > 1000")
# Show sample records from the DataFrame
showDF(highValueTransRecordsFromAPI)
# DataFrame containing bad account records using API
badAccountRecordsFromAPI <- filter(acTransDF, "AccNo NOT like 'SB%'")
# Show sample records from the DataFrame
showDF(badAccountRecordsFromAPI)
# DataFrame containing bad amount records using API
badAmountRecordsFromAPI <- filter(acTransDF, "TranAmount < 0")
# Show sample records from the DataFrame
showDF(badAmountRecordsFromAPI)
# Create a DataFrame by taking the union of two DataFrames
badTransRecordsFromAPI <- union(badAccountRecordsFromAPI, badAmountRecordsFromAPI)
# Show sample records from the DataFrame
showDF(badTransRecordsFromAPI)
# DataFrame containing sum amount using API
sumAmountFromAPI <- agg(goodTransRecordsFromAPI, sumAmount = sum(goodTransRecordsFromAPI$TranAmount))
# Show sample records from the DataFrame
showDF(sumAmountFromAPI)
# DataFrame containing maximum amount using API
maxAmountFromAPI <- agg(goodTransRecordsFromAPI, maxAmount = max(goodTransRecordsFromAPI$TranAmount))
# Show sample records from the DataFrame
showDF(maxAmountFromAPI)
# DataFrame containing minimum amount using API
minAmountFromAPI <- agg(goodTransRecordsFromAPI, minAmount = min(goodTransRecordsFromAPI$TranAmount)) 
# Show sample records from the DataFrame
showDF(minAmountFromAPI)
# DataFrame containing good account number records using API
filteredTransRecordsFromAPI <- filter(goodTransRecordsFromAPI, "AccNo like 'SB%'")
accNosFromAPI <- select(filteredTransRecordsFromAPI, "AccNo")
distinctAccNoFromAPI <- distinct(accNosFromAPI)
sortedAccNoFromAPI <- arrange(distinctAccNoFromAPI, "AccNo")
# Show sample records from the DataFrame
showDF(sortedAccNoFromAPI)
write.parquet(acTransDF, "r.trans.parquet")
acTransDFFromFile <- read.parquet("r.trans.parquet")
# Show sample records from the DataFrame
showDF(acTransDFFromFile)

#Aggregations
# Read data from a JSON file to create DataFrame
acTransDFForAgg <- read.json(paste(DATA_DIR, "TransList2.json", sep = ""))
# Register temporary view definition in the DataFrame for SQL queries
createOrReplaceTempView(acTransDFForAgg, "transnew")
# Show sample records from the DataFrame
showDF(acTransDFForAgg)
# DataFrame containing account summary records using SQL
acSummary <- sql("SELECT AccNo, sum(TranAmount) as TransTotal FROM transnew GROUP BY AccNo")
# Show sample records from the DataFrame
showDF(acSummary)
# DataFrame containing account summary records using API
acSummaryFromAPI <- agg(groupBy(acTransDFForAgg, "AccNo"), TranAmount="sum")
# Show sample records from the DataFrame
showDF(acSummaryFromAPI)

# Multi-data source joins
acMasterDF <- read.json(paste(DATA_DIR, "MasterList.json", sep = ""))
# Show sample records from the DataFrame
showDF(acMasterDF)
# Register temporary view definition in the DataFrame for SQL queries
createOrReplaceTempView(acMasterDF, "master")
acBalDF <- read.json(paste(DATA_DIR, "BalList.json", sep = ""))
# Show sample records from the DataFrame
showDF(acBalDF)
# Register temporary table definition in the DataFrame for SQL queries
createOrReplaceTempView(acBalDF, "balance")
# DataFrame containing account detail records using SQL by joining multiple DataFrame contents
acDetail <- sql("SELECT master.AccNo, FirstName, LastName, BalAmount FROM master, balance WHERE master.AccNo = balance.AccNo ORDER BY BalAmount DESC")
# Show sample records from the DataFrame
showDF(acDetail)
# Persist data in the DataFrame into Parquet file
write.parquet(acDetail, "r.acdetails.parquet")
# Read data into a DataFrame by reading the contents from a Parquet file
acDetailFromFile <- read.parquet("r.acdetails.parquet")
# Show sample records from the DataFrame
showDF(acDetailFromFile)
# Change the column names
acBalDFWithDiffColName <- selectExpr(acBalDF, "AccNo as AccNoBal", "BalAmount")
# Show sample records from the DataFrame
showDF(acBalDFWithDiffColName)
# DataFrame containing account detail records using API by joining multiple DataFrame contents
acDetailFromAPI <- join(acMasterDF, acBalDFWithDiffColName, acMasterDF$AccNo == acBalDFWithDiffColName$AccNoBal)
# Show sample records from the DataFrame
showDF(acDetailFromAPI)
# DataFrame containing account detail records using SQL by selecting specific fields
acDetailFromAPIRequiredFields <- select(acDetailFromAPI, "AccNo", "FirstName", "LastName", "BalAmount")
# Show sample records from the DataFrame
showDF(acDetailFromAPIRequiredFields)

